pkg/apiserver/authentication/identityprovider/oidc/oidc.go

```diff
@@ -0,0 +1,241 @@
+
+	// ClientSecret is the application's secret.
+	ClientSecret string `json:"-" yaml:"clientSecret"`
+
+	// Endpoint contains the resource server's token endpoint
+	// URLs. These are constants specific to each server and are
+	// often available via site-specific packages, such as
+	// google.Endpoint or github.endpoint.
+	Endpoint endpoint `json:"endpoint" yaml:"endpoint"`
+
+	// RedirectURL is the URL to redirect users going through
+	// the OAuth flow, after the resource owner's URLs.
+	RedirectURL string `json:"redirectURL" yaml:"redirectURL"`
+
+	// Scope specifies optional requested permissions.
+	Scopes []string `json:"scopes" yaml:"scopes"`
+
+	// GetUserInfo uses the userinfo endpoint to get additional claims for the token.
+	// This is especially useful where upstreams return "thin" id tokens
+	// https://openid.net/specs/openid-connect-core-1_0.html#UserInfo
+	GetUserInfo bool `json:"getUserInfo" yaml:"getUserInfo"`
+
+	EmailKey             string `json:"emailKey" yaml:"emailKey"`
+	PreferredUsernameKey string `json:"preferredUsernameKey" yaml:"preferredUsernameKey"`
+
+	Provider     *oidc.Provider        `json:"-" yaml:"-"`
+	OAuth2Config *oauth2.Config        `json:"-" yaml:"-"`
+	Verifier     *oidc.IDTokenVerifier `json:"-" yaml:"-"`
+}
+
+// endpoint represents an OAuth 2.0 provider's authorization and token
+// endpoint URLs.
+type endpoint struct {
+	AuthURL     string `json:"authURL" yaml:"authURL"`
+	TokenURL    string `json:"tokenURL" yaml:"tokenURL"`
+	UserInfoURL string `json:"userInfoURL" yaml:"userInfoURL"`
+	JWKSURL     string `json:"jwksURL"`
+}
+
+type oidcIdentity struct {
+	// Subject - Identifier for the End-User at the Issuer.
+	Sub string `json:"sub"`
+	// Shorthand name by which the End-User wishes to be referred to at the RP,
+	// such as janedoe or j.doe. This value MAY be any valid JSON string including special characters such as @, /, or whitespace.
+	// The RP MUST NOT rely upon this value being unique
+	PreferredUsername string `json:"preferred_username"`
+	// End-User's preferred e-mail address.
+	// Its value MUST conform to the RFC 5322 [RFC5322] addr-spec syntax.
+	// The RP MUST NOT rely upon this value being unique.
+	Email string `json:"email"`
+}
+
+type factory struct {
+}
+
+func (f *factory) Type() string {
+	return "OIDCIdentityProvider"
+}
+
+func (f *factory) Create(options oauth.DynamicOptions) (identityprovider.OAuthProvider, error) {
+	var oidcProvider oidcProvider
+	if err := mapstructure.Decode(options, &oidcProvider); err != nil {
+		return nil, err
+	}
+	// dynamically discover
+	if oidcProvider.Issuer != "" {
+		provider, err := oidc.NewProvider(context.TODO(), oidcProvider.Issuer)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create oidc provider: %v", err)
+		}
+		var providerJSON map[string]interface{}
+		if err = provider.Claims(&providerJSON); err != nil {
+			return nil, fmt.Errorf("failed to create oidc provider: %v", err)
```

Comment:
Perhaps you could use a different log output with line [112](https://github.com/kubesphere/kubesphere/pull/3189/files#diff-fb34d9b576729483456ce3d13c2735c6ea75f85588d15d2ffaabd5e4b4c2ccdbR112)

Reply:
By the way, the tests failed. see [here](https://github.com/kubesphere/kubesphere/pull/3189/checks?check_run_id=1628740144#step:7:232)

Reply:
> I left some comments.
> Do we have a OIDC server? Maybe we can use [dex](https://github.com/dexidp/dex/). It has a lot of useful [examples](https://github.com/dexidp/dex/tree/master/examples).

This PR adds the authentication plugin of OIDC identity provider. We will implement the OIDC protocol instead of using dex. Dex can be used as an identity provider.

Reply:
We can use dex to connect to other account systems(as an identity provider), and we can easily convert the dex connector into an authentication plugin.